README.md
[Unforgiving Void](https://minecraft.curseforge.com/projects/unforgiving-void)
============================================================================

[![Curse Forge](http://cf.way2muchnoise.eu/short_274863_downloads.svg)](https://minecraft.curseforge.com/projects/unforgiving-void)

[Issue Reporting](https://github.com/The-Fireplace/Unforgiving-Void/issues)
------------------------------------------------------------------

If you found a bug or even are experiencing a crash please report it so we can fix it. Please check at first if a bug report for the issue already
[exists](https://github.com/The-Fireplace/Unforgiving-Void/issues). If not just create a [new issue](https://github.com/The-Fireplace/Unforgiving-Void/issues/new) and fill out the
form.

Please include the following:

* Minecraft version
* Mod version
* Forge version/build
* Versions of any mods potentially related to the issue 
* Any relevant screenshots/videos are greatly appreciated.
* For crashes:
  * Steps to reproduce
  * fml-client-latest.log or fml-server-latest.log from the logs folder.
 
*(When creating a new issue please follow the template)*

[Feature Requests](https://github.com/The-Fireplace/Unforgiving-Void/issues)
-------------------------------------------------------------------

If you want a new feature added, go ahead an open a [new issue](https://github.com/The-Fireplace/Unforgiving-Void/issues/new), remove the existing form and describe your
feature the best you can. The more details you provide the easier it will be to implement it.
You can also talk to me on [Discord](https://discord.gg/29aj3Ah)

Developing with My Mod
----------------------

If you want to use items or blocks from my mod, add support for, or even develop an addon for my mod, you can easily add it to your development environment! Most
releases get uploaded to my maven repository.  
So all you have to do to include the mod is add these lines *(in the appropriate places)* to your build.gradle

    repositories {
        maven { // The_Fireplace's Mods, and BrainStoneMod
            url "http://maven.brainstonemod.com"
        }
        // Other repos...
    }
    
    dependencies {
        deobfCompile "the_fireplace.unforgivingvoid:UnforgivingVoid:<version>"
        // Other dependencies
    }

Setting up a Workspace/Compiling from Source
--------------------------------------------

* Setup: Run [gradle] in the repository root: `gradlew[.bat] setupDecompWorkspace [eclipse|idea]`
* Build: Run [gradle] in the repository root: `gradlew[.bat] build`
* If obscure Gradle issues are found try running `gradlew clean` and `gradlew cleanCache`(This one is a last resort)


changelog.txt
Updated to 1.15.1


src/main/java/net/blay09/mods/forgivingvoid/ForgivingVoid.java
package net.blay09.mods.forgivingvoid;

import net.minecraft.block.BlockPortal;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.init.MobEffects;
import net.minecraft.init.SoundEvents;
import net.minecraft.item.ItemStack;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.DamageSource;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.common.config.Config;
import net.minecraftforge.event.entity.living.LivingFallEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import net.minecraftforge.fml.relauncher.Side;

import java.util.Random;

@Mod(modid = ForgivingVoid.MOD_ID, name = "Less Forgiving Void", acceptedMinecraftVersions = "[1.10.2]")
@Mod.EventBusSubscriber
public class ForgivingVoid {

	public static final String MOD_ID = "forgivingvoid";

	@SubscribeEvent
	public static void onPlayerTick(TickEvent.PlayerTickEvent event) {
		if(event.side == Side.SERVER && event.phase == TickEvent.Phase.START) {
			if(event.player.posY < 0) {
				event.player.addPotionEffect(new PotionEffect(MobEffects.BLINDNESS, 60, 3));
				Random rand = new Random();
				BlockPos spawnPos = new BlockPos(event.player.posX/8-ModConfig.baseDistanceOffset *2+rand.nextInt(ModConfig.baseDistanceOffset *4), rand.nextInt(100)+16, event.player.posZ/8-ModConfig.baseDistanceOffset *2+rand.nextInt(ModConfig.baseDistanceOffset *4));
				event.player.setPortal(spawnPos);
				if(event.player.changeDimension(-1) != null) {
					int expandMult = 1;
					while(event.player.worldObj.getBlockState(spawnPos).isNormalCube()){
						spawnPos = new BlockPos(event.player.posX/8-ModConfig.baseDistanceOffset *2*expandMult+rand.nextInt(ModConfig.baseDistanceOffset *4*expandMult), rand.nextInt(100)+16, event.player.posZ/8-ModConfig.baseDistanceOffset *2*expandMult+rand.nextInt(ModConfig.baseDistanceOffset *4*expandMult));
						event.player.setPortal(spawnPos);
						expandMult++;
					}
					event.player.setPositionAndUpdate(spawnPos.getX(), spawnPos.getY(), spawnPos.getZ());
					//Even though the player doesn't spawn a portal by doing this, their sudden appearance should destroy any existing portal in the spot they appear.
					for(int x=(int)event.player.posX-2;x<=event.player.posX+2;x++){
						for(int y=(int)event.player.posY-3;y<=event.player.posY+3;y++){
							for(int z=(int)event.player.posZ-2;z<=event.player.posZ+2;z++){
								if(event.player.worldObj.getBlockState(new BlockPos(x, y, z)).getBlock() instanceof BlockPortal) {
									event.player.worldObj.setBlockToAir(new BlockPos(x, y, z));
									event.player.playSound(SoundEvents.BLOCK_GLASS_BREAK, 1.0f, 2.0f);
								}
							}
						}
					}

					if(ModConfig.dropObsidian)
						event.player.worldObj.spawnEntityInWorld(new EntityItem(event.player.worldObj, spawnPos.getX(), spawnPos.getY(), spawnPos.getZ(), new ItemStack(Blocks.OBSIDIAN, 10)));
				}else{
					event.player.setPositionAndUpdate(event.player.posX-ModConfig.baseDistanceOffset *16+rand.nextInt(ModConfig.baseDistanceOffset *32), rand.nextInt(ModConfig.baseDistanceOffset)+ModConfig.baseDistanceOffset /2, event.player.posZ-ModConfig.baseDistanceOffset *16+rand.nextInt(ModConfig.baseDistanceOffset *32));
				}
				event.player.getEntityData().setBoolean("ForgivingVoidNoFallDamage", true);
			}
		}
	}

	@SubscribeEvent
	public static void onPlayerFall(LivingFallEvent event) {
		if(event.getEntity() instanceof EntityPlayer) {
			if(event.getEntity().getEntityData().getBoolean("ForgivingVoidNoFallDamage")) {
				float damage = ModConfig.damageOnFall;
				if(ModConfig.preventDeath && event.getEntityLiving().getHealth() - damage <= 0) {
					damage = event.getEntityLiving().getHealth() - 1f;
				}
				event.getEntity().attackEntityFrom(DamageSource.fall, damage);
				event.setDamageMultiplier(0f);
				event.setCanceled(true);
				event.getEntity().getEntityData().setBoolean("ForgivingVoidNoFallDamage", false);
			}
		}
	}

	@Config(modid = MOD_ID)
	public static class ModConfig {
		@Config.Comment("The amount of damage applied to the player when they land.")
		@Config.RangeInt(min = 0, max = 40)
		public static int damageOnFall = 19;

		@Config.Comment("Prevent death by fall damage when you fall to the nether (limits damage to leave at least 0.5 hearts)")
		public static boolean preventDeath = true;

		@Config.Comment("Drop obsidian for the player when they appear in the Nether")
		public static boolean dropObsidian = true;

		@Config.Comment("The base offset from which the player will spawn from after falling through the void. This is not going to be the exact offset. Height is random.")
		@Config.RangeInt(min = 128, max = 8192)
		public static int baseDistanceOffset = 512;
	}
}


src/main/java/the_fireplace/lessforgivingvoid/LessForgivingVoid.java
package the_fireplace.lessforgivingvoid;

import com.google.common.collect.HashMultimap;
import net.minecraft.block.BlockPortal;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.init.MobEffects;
import net.minecraft.init.SoundEvents;
import net.minecraft.item.ItemStack;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.DamageSource;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.common.config.Config;
import net.minecraftforge.event.entity.living.LivingFallEvent;
import net.minecraftforge.fml.common.*;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import net.minecraftforge.fml.relauncher.Side;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.logging.log4j.Level;

import java.io.File;
import java.util.Random;

@Mod(modid = LessForgivingVoid.MOD_ID, name = "Less Forgiving Void", acceptedMinecraftVersions = "[1.10.2,)", acceptableRemoteVersions = "*", dependencies = "after:forgivingvoid", updateJSON = "https://bitbucket.org/The_Fireplace/minecraft-mod-updates/raw/master/lfv.json")
@Mod.EventBusSubscriber
public class LessForgivingVoid {

	public static final String MOD_ID = "lessforgivingvoid";

	@SubscribeEvent
	public static void onPlayerTick(TickEvent.PlayerTickEvent event) {
		boolean doTeleport = false;
		if(event.side == Side.SERVER && event.phase == TickEvent.Phase.START)
			for(String dimension : ModConfig.dimWhitelist){
				try{
					if(event.player.getEntityWorld().provider.getDimension() == Integer.parseInt(dimension)) {
						doTeleport = true;
						break;
					}
				}catch(NumberFormatException e){
					if(dimension.equals("*")){
						doTeleport = true;
						break;
					}else if(event.player.getEntityWorld().provider.getDimensionType().getName().equals(dimension)){
						doTeleport = true;
						break;
					}
				}
			}

		if(doTeleport) {
			if(event.player.getPosition().getY() < ModConfig.triggerAtY) {
				event.player.addPotionEffect(new PotionEffect(MobEffects.BLINDNESS, 60, 3));
				Random rand = new Random();
				BlockPos spawnPos = new BlockPos(event.player.getPosition().getX()/8-ModConfig.baseDistanceOffset *2+rand.nextInt(ModConfig.baseDistanceOffset *4), rand.nextInt(100)+16, event.player.getPosition().getZ()/8-ModConfig.baseDistanceOffset *2+rand.nextInt(ModConfig.baseDistanceOffset *4));
				event.player.setPortal(spawnPos);
				if(event.player.changeDimension(-1) != null) {
					int expandMult = 1;
					while(event.player.world.getBlockState(spawnPos).isNormalCube()){
						spawnPos = new BlockPos(event.player.getPosition().getX()/8-ModConfig.baseDistanceOffset *2*expandMult+rand.nextInt(ModConfig.baseDistanceOffset *4*expandMult), rand.nextInt(100)+16, event.player.getPosition().getZ()/8-ModConfig.baseDistanceOffset *2*expandMult+rand.nextInt(ModConfig.baseDistanceOffset *4*expandMult));
						event.player.setPortal(spawnPos);
						expandMult++;
					}
					event.player.setPositionAndUpdate(spawnPos.getX(), spawnPos.getY(), spawnPos.getZ());
					//Even though the player doesn't spawn a portal by doing this, their sudden appearance should destroy any existing portal in the spot they appear.
					for(int x=event.player.getPosition().getX()-2;x<=event.player.getPosition().getX()+2;x++){
						for(int y=event.player.getPosition().getY()-3;y<=event.player.getPosition().getY()+3;y++){
							for(int z=event.player.getPosition().getZ()-2;z<=event.player.getPosition().getZ()+2;z++){
								if(event.player.world.getBlockState(new BlockPos(x, y, z)).getBlock() instanceof BlockPortal) {
									event.player.world.setBlockToAir(new BlockPos(x, y, z));
									event.player.playSound(SoundEvents.BLOCK_GLASS_BREAK, 1.0f, 2.0f);
								}
							}
						}
					}

					if(ModConfig.dropObsidian)
						event.player.world.spawnEntity(new EntityItem(event.player.world, spawnPos.getX(), spawnPos.getY(), spawnPos.getZ(), new ItemStack(Blocks.OBSIDIAN, 10)));
					event.player.getEntityData().setBoolean("ForgivingVoidNoFallDamage", true);
				}else if(!event.player.world.isRemote && !event.player.isDead)
					FMLLog.log(Level.WARN, "Error: Unable to teleport player to the Nether from the void. Unfortunately, the player will die. If this happens, please report it.");
			}
		}
	}

	@SubscribeEvent
	public static void onPlayerFall(LivingFallEvent event) {
		if(event.getEntity() instanceof EntityPlayer) {
			if(event.getEntity().getEntityData().getBoolean("ForgivingVoidNoFallDamage")) {
				float damage = ModConfig.damageOnFall;
				if(ModConfig.preventDeath && event.getEntityLiving().getHealth() - damage <= 0) {
					damage = event.getEntityLiving().getHealth() - 1f;
				}
				event.getEntity().attackEntityFrom(DamageSource.FALL, damage);
				event.setDamageMultiplier(0f);
				event.setCanceled(true);
				event.getEntity().getEntityData().setBoolean("ForgivingVoidNoFallDamage", false);
			}
		}
	}

	@Config(modid = MOD_ID)
	public static class ModConfig {
		@Config.Comment("The y level at which Less Forgiving Void should send the player to the Nether.")
		@Config.RangeInt(min = -64, max = 0)
		public static int triggerAtY = -32;

		@Config.Comment("The amount of damage applied to the player when they land.")
		@Config.RangeInt(min = 0, max = 40)
		public static int damageOnFall = 19;

		@Config.Comment("Prevent death by fall damage when you fall to the nether (limits damage to leave at least 0.5 hearts)")
		public static boolean preventDeath = true;

		@Config.Comment("Drop obsidian for the player when they appear in the Nether")
		public static boolean dropObsidian = true;

		@Config.Comment("The base offset from which the player will spawn from after falling through the void. This is not going to be the exact offset. Height is random.")
		@Config.RangeInt(min = 128, max = 8192)
		public static int baseDistanceOffset = 512;

		@Config.Comment("The dimensions the mod is enabled in. Include a * to make the mod work in all dimensions. This can use dimension name or ID Number")
		public static String[] dimWhitelist = new String[]{"*"};
	}
}


src/main/java/the_fireplace/unforgivingvoid/UnforgivingVoidCapability.java
package the_fireplace.unforgivingvoid;

import net.minecraft.nbt.ByteNBT;
import net.minecraft.nbt.INBT;
import net.minecraft.util.Direction;
import net.minecraftforge.common.capabilities.Capability;

import javax.annotation.Nullable;

public interface UnforgivingVoidCapability {

    boolean getFallingFromTravel();

    void setFallingFromTravel(boolean fallingFromTravel);

    class Default implements UnforgivingVoidCapability {
        private boolean isFallingFromTravel;

        public Default() {
            isFallingFromTravel = false;
        }

        @Override
        public void setFallingFromTravel(boolean fallingFromTravel) {
            isFallingFromTravel = fallingFromTravel;
        }

        @Override
        public boolean getFallingFromTravel() {
            return isFallingFromTravel;
        }
    }

    class Storage implements Capability.IStorage<UnforgivingVoidCapability> {

        @Nullable
        @Override
        public INBT writeNBT(Capability<UnforgivingVoidCapability> capability, UnforgivingVoidCapability instance, Direction side) {
            return instance.getFallingFromTravel() ? ByteNBT.field_229670_c_ : ByteNBT.field_229669_b_;
        }

        @Override
        public void readNBT(Capability<UnforgivingVoidCapability> capability, UnforgivingVoidCapability instance, Direction side, INBT nbt) {
            if (nbt instanceof ByteNBT)
                instance.setFallingFromTravel(((ByteNBT) nbt).getByte() == 1);
        }
    }
}

src/main/java/the_fireplace/unforgivingvoid/UnforgivingVoid.java
package the_fireplace.unforgivingvoid;

import com.google.common.collect.Lists;
import mcp.MethodsReturnNonnullByDefault;
import net.minecraft.block.AirBlock;
import net.minecraft.block.Blocks;
import net.minecraft.block.NetherPortalBlock;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.ItemEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.INBT;
import net.minecraft.potion.EffectInstance;
import net.minecraft.potion.Effects;
import net.minecraft.util.DamageSource;
import net.minecraft.util.Direction;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.SoundEvents;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.dimension.DimensionType;
import net.minecraftforge.common.ForgeConfigSpec;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.CapabilityManager;
import net.minecraftforge.common.capabilities.ICapabilitySerializable;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.event.AttachCapabilitiesEvent;
import net.minecraftforge.event.TickEvent;
import net.minecraftforge.event.entity.living.LivingFallEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.LogicalSide;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.config.ModConfig;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.apache.commons.lang3.tuple.Pair;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.List;
import java.util.Objects;
import java.util.Random;

@SuppressWarnings("WeakerAccess")
@MethodsReturnNonnullByDefault
@Mod(UnforgivingVoid.MODID)
public class UnforgivingVoid {
	public static final String MODID = "unforgivingvoid";
	private static Logger LOGGER = LogManager.getLogger(MODID);

	@CapabilityInject(UnforgivingVoidCapability.class)
	public static final Capability<UnforgivingVoidCapability> UNFORGIVING_VOID_CAP = null;
	private static final ResourceLocation unforgiving_void_cap_res = new ResourceLocation(MODID, "unforgiving_void_cap");

	public UnforgivingVoid() {
		ModLoadingContext.get().registerConfig(ModConfig.Type.COMMON, cfg.COMMON_SPEC);
		FMLJavaModLoadingContext.get().getModEventBus().addListener(this::preInit);
		FMLJavaModLoadingContext.get().getModEventBus().addListener(this::serverConfig);
	}

	@SuppressWarnings("unused")
	public void preInit(FMLCommonSetupEvent event) {
		CapabilityManager.INSTANCE.register(UnforgivingVoidCapability.class, new UnforgivingVoidCapability.Storage(), UnforgivingVoidCapability.Default::new);
		MinecraftForge.EVENT_BUS.register(this);
	}

	public void serverConfig(ModConfig.ModConfigEvent event) {
		if (event.getConfig().getType() == ModConfig.Type.COMMON)
			cfg.load();
	}

	@SubscribeEvent
	public void attachPlayerCaps(AttachCapabilitiesEvent<Entity> e){
		if(e.getObject() instanceof PlayerEntity) {
			//noinspection ConstantConditions
			assert UNFORGIVING_VOID_CAP != null;
			e.addCapability(unforgiving_void_cap_res, new ICapabilitySerializable() {
				UnforgivingVoidCapability inst = UNFORGIVING_VOID_CAP.getDefaultInstance();

				@Override
				public INBT serializeNBT() {
					return UNFORGIVING_VOID_CAP.getStorage().writeNBT(UNFORGIVING_VOID_CAP, inst, null);
				}

				@Override
				public void deserializeNBT(INBT nbt) {
					UNFORGIVING_VOID_CAP.getStorage().readNBT(UNFORGIVING_VOID_CAP, inst, null, nbt);
				}

				@Override
				public <T> LazyOptional<T> getCapability(@Nonnull Capability<T> capability, @Nullable Direction facing) {
					//noinspection unchecked
					return capability.getName().equals(UnforgivingVoidCapability.class.getCanonicalName()) ? LazyOptional.of(() -> (T) inst) : LazyOptional.empty();
				}
			});
		}
	}

	@SubscribeEvent
	public void onPlayerTick(TickEvent.PlayerTickEvent event) {
		boolean doTeleport = cfg.dimensionFilter.contains("*");
		if(event.side == LogicalSide.SERVER && event.phase == TickEvent.Phase.START)
			for (String dim : cfg.dimensionFilter)
				try{
					if(event.player.getEntityWorld().getDimension().getType().getId() == Integer.parseInt(dim)) {
						doTeleport = !doTeleport;
						break;
					}
				}catch(NumberFormatException e){
					if(!dim.equals("*") && (Objects.requireNonNull(event.player.getEntityWorld().getDimension().getType().getRegistryName()).toString().toLowerCase().equals(dim.toLowerCase()) || event.player.getEntityWorld().getDimension().getType().getRegistryName().getPath().toLowerCase().equals(dim.toLowerCase()))){
						doTeleport = !doTeleport;
						break;
					}
				}

		if(doTeleport && event.player.getPosition().getY() <= cfg.triggerAtY) {
			event.player.addPotionEffect(new EffectInstance(Effects.BLINDNESS, 60, 3));
			Random rand = new Random();
			BlockPos spawnPos = new BlockPos(event.player.getPosition().getX()/8 - cfg.baseDistanceOffset + rand.nextInt(cfg.baseDistanceOffset * 2), rand.nextInt(100)+16, event.player.getPosition().getZ()/8 - cfg.baseDistanceOffset + rand.nextInt(cfg.baseDistanceOffset * 2));
			event.player.setPortal(spawnPos);
			if(event.player.changeDimension(DimensionType.THE_NETHER) != null && event.player.dimension.equals(DimensionType.THE_NETHER)) {
				int expandMult = 1;
				while(event.player.world.getBlockState(spawnPos).isNormalCube(event.player.world, spawnPos)){
					spawnPos = new BlockPos(event.player.getPosition().getX()/8 - cfg.baseDistanceOffset * expandMult + rand.nextInt(cfg.baseDistanceOffset * 2 * expandMult), rand.nextInt(100)+16, event.player.getPosition().getZ()/8 - cfg.baseDistanceOffset * expandMult + rand.nextInt(cfg.baseDistanceOffset * 2 * expandMult));
					event.player.setPortal(spawnPos);
					expandMult++;
				}
				event.player.setPositionAndUpdate(spawnPos.getX(), spawnPos.getY(), spawnPos.getZ());
				//Even though the player doesn't spawn a portal by doing this, their sudden appearance should destroy any existing portal in the spot they appear.
				for(int x=event.player.getPosition().getX()-2;x<=event.player.getPosition().getX()+2;x++){
					for(int y=event.player.getPosition().getY()-3;y<=event.player.getPosition().getY()+3;y++){
						for(int z=event.player.getPosition().getZ()-2;z<=event.player.getPosition().getZ()+2;z++){
							if(event.player.world.getBlockState(new BlockPos(x, y, z)).getBlock() instanceof NetherPortalBlock) {
								event.player.world.setBlockState(new BlockPos(x, y, z), Blocks.AIR.getDefaultState());
								event.player.playSound(SoundEvents.BLOCK_GLASS_BREAK, 1.0f, 2.0f);
							}
						}
					}
				}

				if(cfg.dropObsidian)
					event.player.world.addEntity(new ItemEntity(event.player.world, spawnPos.getX(), spawnPos.getY(), spawnPos.getZ(), new ItemStack(Blocks.OBSIDIAN, 10+rand.nextInt(4))));
				if(!event.player.isCreative() && !event.player.isSpectator())
					getUnforgivingVoidCap(event.player).setFallingFromTravel(true);

				if(cfg.lavaPlatform) {
					while (event.player.world.getBlockState(spawnPos).getBlock() instanceof AirBlock)
						spawnPos = spawnPos.down();
					if (event.player.world.getBlockState(spawnPos).getBlock() == Blocks.LAVA) {
						event.player.getEntityWorld().setBlockState(spawnPos, Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.north(), Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.south(), Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.east(), Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.west(), Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.north().east(), Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.north().west(), Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.south().east(), Blocks.OBSIDIAN.getDefaultState());
						event.player.getEntityWorld().setBlockState(spawnPos.south().west(), Blocks.OBSIDIAN.getDefaultState());
					}
				}
			} else if(!event.player.world.isRemote && event.player.isAlive())
				LOGGER.warn("Error: Unable to teleport player to the Nether from the void. Unfortunately, the player will die. If this happens, please report it.");
		}
	}

	public static UnforgivingVoidCapability getUnforgivingVoidCap(PlayerEntity player) {
		//noinspection ConstantConditions
		return player.getCapability(UNFORGIVING_VOID_CAP).orElseThrow(() -> new IllegalStateException("Unforgiving Void Capability is not present for a player!"));
	}

	@SubscribeEvent
	public void onPlayerFall(LivingFallEvent event) {
		if(event.getEntityLiving() instanceof PlayerEntity && event.getEntityLiving().isServerWorld()) {
			if(getUnforgivingVoidCap((PlayerEntity) event.getEntityLiving()).getFallingFromTravel()) {
				float damage = cfg.damageOnFall;
				if(cfg.preventFallingDeath && event.getEntityLiving().getHealth() - damage <= 0)
					damage = event.getEntityLiving().getHealth() - 1f;
				event.getEntity().attackEntityFrom(DamageSource.FALL, damage);
				event.setDamageMultiplier(0f);
				event.setCanceled(true);
				getUnforgivingVoidCap((PlayerEntity) event.getEntity()).setFallingFromTravel(false);
			}
		}
	}

	public static class cfg {
		public static final ServerConfig COMMON;
		public static final ForgeConfigSpec COMMON_SPEC;
		static {
			final Pair<ServerConfig, ForgeConfigSpec> specPair = new ForgeConfigSpec.Builder().configure(ServerConfig::new);
			COMMON_SPEC = specPair.getRight();
			COMMON = specPair.getLeft();
		}

		public static int triggerAtY;
		public static int damageOnFall;
		public static boolean preventFallingDeath;
		public static boolean dropObsidian;
		public static boolean lavaPlatform;
		public static int baseDistanceOffset;
		public static List<String> dimensionFilter;

		public static void load() {
			triggerAtY = COMMON.triggerAtY.get();
			damageOnFall = COMMON.damageOnFall.get();
			preventFallingDeath = COMMON.preventFallingDeath.get();
			dropObsidian = COMMON.dropObsidian.get();
			lavaPlatform = COMMON.lavaPlatform.get();
			baseDistanceOffset = COMMON.baseDistanceOffset.get();
			dimensionFilter = COMMON.dimensionFilter.get();
		}

		public static class ServerConfig {
			//General clan config
			public ForgeConfigSpec.IntValue triggerAtY;
			public ForgeConfigSpec.IntValue damageOnFall;
			public ForgeConfigSpec.BooleanValue preventFallingDeath;
			public ForgeConfigSpec.BooleanValue dropObsidian;
			public ForgeConfigSpec.BooleanValue lavaPlatform;
			public ForgeConfigSpec.IntValue baseDistanceOffset;
			public ForgeConfigSpec.ConfigValue<List<String>> dimensionFilter;

			ServerConfig(ForgeConfigSpec.Builder builder) {
				builder.push("general");
				triggerAtY = builder
						.comment("The y level at which Unforgiving Void should send the player to the Nether.")
						.translation("Trigger At Y")
						.defineInRange("triggerAtY", -32, Integer.MIN_VALUE, Integer.MAX_VALUE);
				damageOnFall = builder
						.comment("The amount of damage applied to the player when they land in the Nether.")
						.translation("Damage On Fall")
						.defineInRange("damageOnFall", 19, 0, Integer.MAX_VALUE);
				preventFallingDeath = builder
						.comment("Prevent death by fall damage when you fall to the Nether (limits damage to leave at least 0.5 hearts)")
						.translation("Prevent Falling Death")
						.define("preventFallingDeath", true);
				dropObsidian = builder
						.comment("Drop obsidian for the player when they appear in the Nether.")
						.translation("Drop Obsidian")
						.define("dropObsidian", true);
				lavaPlatform = builder
						.comment("Attempt to place a platform for the player when they appear in the Nether above lava.")
						.translation("Lava Platform")
						.define("lavaPlatform", false);
				baseDistanceOffset = builder
						.comment("The base offset from which the player will spawn from after falling through the void. This is not going to be the exact offset. Height is random.")
						.translation("Base Distance Offset")
						.defineInRange("baseDistanceOffset", 512, 1, Integer.MAX_VALUE);
				dimensionFilter = builder
						.comment("This is the Dimension Filter. It specifies which dimensions Unforgiving Void is active in. If it contains *, it is a blacklist. Otherwise, it is a whitelist.")
						.translation("Dimension Filter")
						.define("dimensionFilter", Lists.newArrayList("*"));
				builder.pop();
			}
		}
	}
}
